src/CoreWCF.RabbitMQ.Client/src/CoreWCF/ServiceModel/Channels/RabbitMqOutputChannel.cs

```diff
@@ -0,0 +1,284 @@
+
+        public void Send(Message message)
+        {
+            SendAsync(message).Wait();
+        }
+
+        public void Send(Message message, TimeSpan timeout)
+        {
+            SendAsync(message, timeout).Wait();
+        }
+
+        public IAsyncResult BeginSend(Message message, AsyncCallback callback, object state)
+        {
+            return SendAsync(message).ToApm(callback, state);
+        }
+
+        public IAsyncResult BeginSend(Message message, TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            return SendAsync(message, timeout).ToApm(callback, state);
+        }
+
+        public void EndSend(IAsyncResult result)
+        {
+            result.ToApmEnd();
+        }
+
+        /// <summary>
+        /// Address the Message and serialize it into a byte array.
+        /// </summary>
+        private ArraySegment<byte> EncodeMessage(Message message)
+        {
+            try
+            {
+                _baseAddress.ApplyTo(message);
+                return _encoder.WriteMessage(
+                    message,
+                    (int)_parent.Transport.MaxReceivedMessageSize,
+                    _parent.BufferManager);
+            }
+            finally
+            {
+                // We have consumed the message by serializing it, so clean up
+                message.Close();
+            }
+        }
+
+        private async Task OnOpenAsync(TimeSpan timeout)
+        {
+            string userName = ConnectionFactory.DefaultUser;
+            string password = ConnectionFactory.DefaultPass;
+
+            var token = await GetSecurityTokenAsync(timeout);
+            if (token != null)
+            {
+                // When UserNameSecurityToken is made public, cast token to UserNameSecurityToken
+                // and read UserName and Password properties directly instead of using reflection.
+                userName = GetPropertyValueFromReflection(token, "UserName");
+                password = GetPropertyValueFromReflection(token, "Password");
+            }
+
+            _connectionSettings = GetConnectionSettings(userName, password);
+            var factory = _connectionSettings.GetConnectionFactory();
+            var connection = factory.CreateConnection();
+            _rabbitMqClient = connection.CreateModel();
+            _rabbitMqClient.ConfirmSelect();
+        }
+
+        private string GetPropertyValueFromReflection(SecurityToken token, string propertyName)
+        {
+            var tokenType = token.GetType();
+            var propertyValue = tokenType.GetProperty(propertyName)?.GetValue(token, null);
+            return propertyValue as string ?? string.Empty;
+        }
+
+        /// <summary>
+        /// Published a Message to RabbitMQ.
+        /// </summary>
+        /// <exception cref="TimeoutException"></exception>
+        private Task SendAsync(Message message)
+        {
+            return SendAsync(message, TimeSpan.Zero);
+        }
```

Comment:
TimeSpan.Zero generally means execute the functionality you are asking for, but don't wait if it doesn't complete immediately. With many api's, passing Zero will throw. The value you want to pass is `System.Threading.Timeout.InfiniteTimeSpan`